invoices/invoiceregistry.go

```diff
@@ -95,21 +99,51 @@ func NewRegistry(cdb *channeldb.DB, finalCltvRejectDelta int32) *InvoiceRegistry
 		hodlSubscriptions:         make(map[channeldb.CircuitKey]map[chan<- interface{}]struct{}),
 		hodlReverseSubscriptions:  make(map[chan<- interface{}]map[channeldb.CircuitKey]struct{}),
 		finalCltvRejectDelta:      finalCltvRejectDelta,
+		expiryWatcher:             expiryWatcher,
 		quit:                      make(chan struct{}),
 	}
 }
 
+// prefetchPending fetches all active invoices and their corresponding payment
+// hashes from ChannelDB and adds them to the expiry watcher.
+func (i *InvoiceRegistry) prefetchPendingInvoices() error {
+	pendingOnly := true
+	pendingInvoices, err := i.cdb.FetchAllInvoicesWithPaymentHash(pendingOnly)
+	if err != nil && err != channeldb.ErrNoInvoicesCreated {
+		log.Errorf(
+			"Error while prefetching active invoices from the database: %v", err,
+		)
+		return err
+	}
+
+	for idx := range pendingInvoices {
+		i.expiryWatcher.AddInvoice(
+			pendingInvoices[idx].PaymentHash, &pendingInvoices[idx].Invoice,
+		)
+	}
+
+	return nil
+}
+
 // Start starts the registry and all goroutines it needs to carry out its task.
 func (i *InvoiceRegistry) Start() error {
-	i.wg.Add(1)
+	// Start InvoiceExpiryWatcher and prepopulate it with existing active
+	// invoices.
+	if err := i.expiryWatcher.Start(i.CancelInvoice); err != nil {
+		return err
+	}
 
+	i.wg.Add(1)
 	go i.invoiceEventNotifier()
 
-	return nil
+	// Now prefetch all pending invoices to the expiry watcher.
+	return i.prefetchPendingInvoices()
```

Comment:
Is this the right order? Start go routine before prefetch?

Reply:
Yes, `InvoiceExpiryWatcher` must be started before `InvoiceExpiryWatcher.AddInvoice` may be called, otherwise it'll block until it is not started.